engine/access/rpc/backend/transactions/provider/execution_node.go

```diff
@@ -251,27 +295,46 @@ func (e *ENTransactionProvider) TransactionResultsByBlockID(
 		}
 	}
 
-	// after iterating through all transactions  in each collection, i equals the total number of
-	// user transactions  in the block
-	txCount := i
-	sporkRootBlockHeight := e.state.Params().SporkRootBlockHeight()
+	return results, nil
+}
 
-	// root block has no system transaction result
-	if block.Height > sporkRootBlockHeight {
-		// system chunk transaction
+// systemTransactionResults constructs the system transaction results from the execution node response.
+//
+// It does so by itterating through all system transactions in the block and constructing the transaction results.
+// System transactions are transactions that follow the user transactions from the execution node response.
+// We should always return transaction result for system chunk transaction, but if scheduled callbacks are enabled
+// we also return results for the process and execute callbacks transactions.
+func (e *ENTransactionProvider) systemTransactionResults(
+	ctx context.Context,
+	userTxCount int,
+	block *flow.Block,
+	blockID flow.Identifier,
+	resp *execproto.GetTransactionResultsResponse,
+	requiredEventEncodingVersion entities.EventEncodingVersion,
+) ([]*accessmodel.TransactionResult, error) {
 
-		// resp.TransactionResultsByBlockID includes the system tx result, so there should be exactly one
-		// more result than txCount
-		if txCount != len(resp.TransactionResults)-1 {
-			if txCount >= len(resp.TransactionResults) {
-				return nil, errInsufficientResults
-			}
-			// otherwise there are extra results
-			// TODO(bft): slashable offense
-			return nil, status.Errorf(codes.Internal, "number of transaction results returned by execution node is more than the number of transactions  in the block")
-		}
+	// resp.TransactionResultsByBlockID includes the system tx results plus the user transactions
+	allTxCount := len(resp.TransactionResults)
+	systemTxCount := allTxCount - userTxCount
+	systemTxStartIndex := allTxCount - systemTxCount
+	// should never happen
+	if systemTxCount <= 0 {
+		return nil, status.Errorf(codes.Internal, "no system transaction results")
+	}
```

Comment:
I think this logic makes the checks clearer
```suggestion
	if userTxCount >= len(resp.TransactionResults) {
		return nil, status.Errorf(codes.Internal, "no system transaction results")
	}

	systemTxStartIndex := userTxCount
	systemTxResults := resp.TransactionResults[systemTxStartIndex:]
	systemTxCount := len(systemTxResults)
```

this also replaces line 325 below, the UI just wouldn't let me include it